### PR TITLE
New Diaspora implementation

### PR DESCRIPTION
**Pull request by annando**
This is a complete rework of the Diaspora protocol. Everything should be checked. This will enable us to keep in touch with the upcoming changes in the protocol. This implementation should be able to process the old and the proposed new format.

### CODE REVIEW

File: include/xml.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
+<?php
```
Comment (rabuzarus) [inline include/xml.php]:
```
/**
 * @file include/xml.php
 */

**Comment by rabuzarus** [comment]
short comment/question
- include/xml.php needs some doxygen docu for it's methods

I have one problem. For include/diaspora.php github says 
>2,608 additions, 2,566 deletions not shown because the diff is too large. Please use a local Git client to view these changes.

 Is there a possibility to comment on the lines? 

At the moment I'm very busy. I will look through the the PR from time to time. But because it is very much code, it will take some time

**Comment by annando** [comment]
I don't know if commenting on that lines is possible. Good thing is: This is the second largest file that we are having - so this won't happen too much in the future :-) If you want to have look, you can do it here: https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php

**Comment by rabuzarus** [comment]
- `@return` misses the "type" of the variable - string (https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L119)
- please use `/*` instead of `/**` because otherwise it would be recognized as doxygen (https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L182)
- maybe some explanation in the line below for what the subfunction is needed (https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L798)
- https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L938
```
 * @return array
 *      'cid' => contact id
 *      'network' => network type
```

More to come ;-)

**Comment by rabuzarus** [comment]
OK, some more stuff ;-)
- I would leave https://github.com/friendica/friendica/pull/2418/commits/c284ab5eff79fa2a87f7214fe433001712db08c9#diff-e49823869ea50948ade5b4d497c64b9fL182 inside because it is a nice example how a decrypted header looks like - (just remove one `*` in the first line to have only `/*` so it would not be in the doxygen)
- `arr['object']` (https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L1697) should constructed similar like it's done here  https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L1326 to make it more readable
- I don't know if https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L2187 would appear in doxygen. I haven't tested it but I think so. It should be changed to:
```
	/*
	 * Here are all the functions that are needed to transmit data with the Diaspora protocol *
	 ******************************************************************************************/
```
- function store_like_signature & function store_comment_signature should be defined as `private` or `public static`
- maybe some examples like in https://github.com/friendica/friendica/pull/2418/commits/c284ab5eff79fa2a87f7214fe433001712db08c9#diff-e49823869ea50948ade5b4d497c64b9fL182 would be also nice at other places to show newcomers and interested persons how some parts of the diaspora message look like and what we do with it

**Comment by annando** [comment]
@rabuzarus I did some of the changes you suggested. The examples doesn't make much sense since the format will change soon in the future.

File: include/xml.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ */
+class xml {
+	/**
+	 * @brief Creates an XML structure out of a given array
+	 *
+	 * @param array $array The array of the XML structure that will be generated
+	 * @param object $xml The createdXML will be returned by reference
+	 * @param bool $remove_header Should the XML header be removed or not?
+	 * @param array $namespaces List of namespaces
```
Comment (rabuzarus) [inline include/xml.php]:
```
 * @param bool $root
```

File: util/createdoxygen.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+		if ($found) {
+			echo add_documentation($line);
+		}
+	}
+	echo $line."\n";
+	$previous = $line;
+}
+
+function add_documentation($line) {
```
Comment (rabuzarus) [inline util/createdoxygen.php]: add some doxygen docu

**Comment by rabuzarus** [comment]
- delete the `/` in https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L758
- https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L2145 :
```
// Checking for embedded pictures
```
(taken from https://github.com/friendica/friendica/blob/develop/include/diaspora.php#L1081)
- I would keep `// see: https://github.com/Raven24/diaspora-federation/blob/master/lib/diaspora-federation/entities/reshare.rb` at https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L2637

> The examples doesn't make much sense since the format will change soon in the future
I know that they will do some changes to the diaspora protocol. But will friendicas support for old diaspora server cease with this step? 
Maybe we should add a `@todo: Give some examples`

There was also a paragraphp `How Diaspora performs comment signature checking:`, `How Diaspora performs "like" signature checking:` and `How Diaspora performs relayable_retraction signature checking:` (https://github.com/friendica/friendica/blob/develop/include/diaspora.php#L1459 , https://github.com/friendica/friendica/blob/develop/include/diaspora.php#L2049 and https://github.com/friendica/friendica/blob/develop/include/diaspora.php#L2284). Is this also obsolete? If so we should rather update it than deleting it.

Does diaspora have a docu for it's protocol? If so we could insert a link at the header of the file.

**Comment by annando** [comment]
The raven repository is outdated and many of this stuff will change. The explanation concerning the signature check was outdated as well.

I'm more with "@todo: give some examples" ;-)

I'm unsure if we really should document a foreign protocol. I hope that @SuperTux88 will do the documentation. Then we could link to it. Maybe we could set a link to https://github.com/SuperTux88/diaspora_federation

**Comment by SuperTux88** [comment]
Yes, I'll write a complete protocol documentation/specification once everything is finished. I created issues with all planned changes here: https://github.com/SuperTux88/diaspora_federation/issues They are currently still open for feedback/changes.

**Comment by annando** [comment]
Concerning the support for old Diaspora servers: We will do it according to Diaspora as well. That means the following steps:

1. We are able to receive the old and the new format, but we will send the old format. (This will be possible with this pull request)
2. After some time we will send the new format. Then we will only accept the new format.

We could - however - add some version checking to the protocol part. But this would increase complexity. And since the Diaspora part is really huge by now, I would like to reduce complexity - and not add it.

BTW: This implemention here is based on the unfinished draft. In fact we are faster with the implementation than Diaspora :-) That means that in the future we could have some small changes to really be able to accept the new format.

**Comment by SuperTux88** [comment]
The second step should be two steps, so:
1. receive old and new, send old
2. still receive old and new, send new
3. remove support for old stuff and cleanup

Between the steps should be 4-6 Month (starting with the 0.6.0.0 release of diaspora). I think we don't need the complexity of version-checks. If people don't update their server in the 4-6 Month, they are out.

I don't expect many changes now, because most changes are done now. :smiley:

**Comment by annando** [comment]
You are right - I forgot one step.

**Comment by rabuzarus** [comment]
- doesn't `$fields` have to be defined? (https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L326)
- The `@return` is wrong - It is is an array which contains the whole contact data, not only the ID (relayable_retraction)
- is this always bool (https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L740). Just asking because I don't know if in this case `$contact = self::contact_by_handle($importer["uid"], $handle);` would always give a type: bool
- https://github.com/annando/friendica/blob/1602-diaspora/include/diaspora.php#L934
```
 * @return array
 *      'cid' => contact id
 *      'network' => network type
```
- have a look at the inline comments above for util/createdoxygen.php and include/xml.php

**Comment by rabuzarus** [comment]
+
- include/xml.php: The class xml needs a `@brief` description. The actual `@brief` is used as file description

**Comment by rabuzarus** [comment]
+
- include/diaspora.php: function repair_signature & function decode should be defined as private or public static + the same should be done for methods in include/xml.php

**Comment by annando** [comment]
- I don't think that $fields needs a definition since it is a call by reference
- xml.php does have this "brief" - I don't understand what you mean
- What do you mean with "have a look at the inline comments above for util/createdoxygen.php and include/xml.php"?

**Comment by annando** [comment]
To which line do you refer with the comment "The return is wrong"?

**Comment by annando** [comment]
The rest of the work must wait until I'm at home again. (Tomorrow evening)

**Comment by rabuzarus** [comment]
> xml.php does have this "brief" - I don't understand what you mean

Yes but the `@brief` below `@file` is used by doxygen as file description. There have to be another inside xml.php as class description

**Comment by annando** [comment]
@rabuzarus Is that correct now?

**Comment by rabuzarus** [comment]
All is fine from my side. I have nothing more to add

**Comment by annando** [comment]
@rabuzarus Uih ... :-)

**Comment by rabuzarus** [comment]
Good work

**Pull request merged**